packages/rpc-methods/src/request.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  PermissionSpecificationBuilder,
+  PermissionType,
+  RestrictedMethodOptions,
+} from '@metamask/permission-controller';
+import { PermittedHandlerExport } from '@metamask/types';
+import { JsonRpcParams } from '@metamask/utils';
+
+import { methodHandlers } from './permitted';
+import { restrictedMethodPermissionBuilders } from './restricted';
+
+/**
+ * Get the method implementation from a {@link PermittedHandlerExport}.
+ *
+ * @template Handler - A permitted handler export.
+ */
+type PermittedMethodImplementation<Handler> =
+  Handler extends PermittedHandlerExport<any, infer Args, infer Result>
+    ? (args: Args) => Promise<Result>
+    : never;
+
+/**
+ * Get a JSON-RPC method type from a {@link PermittedHandlerExport} and a method
+ * name.
+ *
+ * @template MethodName - The name of the method.
+ * @template Handler - A permitted handler export.
+ */
+type PermittedMethod<
+  MethodName extends string,
+  Handler,
+> = PermittedMethodImplementation<Handler> extends (
+  args: infer Args,
+) => infer Return
+  ? (args: { method: MethodName; params?: Args }) => Return
+  : never;
+
+/**
+ * Get a restricted method implementation from a
+ * {@link PermissionSpecificationBuilder}.
+ *
+ * @template Builder - A permission specification builder.
+ */
+type RestrictedMethodImplementation<Builder> = Builder extends {
+  specificationBuilder: PermissionSpecificationBuilder<
+    PermissionType.RestrictedMethod,
+    any,
+    infer Specification
+  >;
+}
+  ? Specification['methodImplementation']
+  : never;
+
+/**
+ * Get a JSON-RPC method type from a {@link PermissionSpecificationBuilder}.
+ *
+ * @template Builder - A permission specification builder.
+ */
+type RestrictedMethod<Builder extends { targetKey: string }> =
+  RestrictedMethodImplementation<Builder> extends (
+    args: infer Args,
+  ) => infer Return
+    ? Args extends RestrictedMethodOptions<infer Params>
+      ? (args: { method: Builder['targetKey']; params?: Params }) => Return
+      : never
+    : never;
+
+/**
+ * A type containing all permitted JSON-RPC methods.
+ */
+type PermittedMethodFunction = {
```

Comment:
Nitpicking but using `Function` and `Arguments` suffixes for types for JSON-RPC methods rather than TypeScript functions is confusing